gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h

```diff
@@ -111,13 +142,21 @@ struct _GstTensorTrainerFramework
    * @param[in] prop read-only property values
    * @param[in] private_data A subplugin may save its internal private data here.
    * @param[out] fw_info struct to hold frameworks info. Must be allocated by the caller (return value).
-   * @return 0 if OK. non-zero if error.
+   * @return 0 if OK. < 0 if error.
    *
    * @note CAUTION: private_data can be NULL if the framework is not yet opened by the caller.
    */
 
-  /* Need to make (*eventHandler)*/
-
+  int (*eventHandler) (const GstTensorTrainerFramework *self,
```

Comment:
Checking completions of epoch or training-session is NOT an event. It's rather "getStatus".
And... if epoch1 is done and it's running epoch2, what should it return for "getStatus(isEpochComplete)"? It's "TRUE" (epoch1 is completed) or "FALSE" (epoch2 is not completed)?

Reply:
The current implementation checks whether the epoch is complete when the data required for 1 epoch have been pushed in the chain function.

The check method calls eventHandler and checks whether is_epoch_complete of eventData received from sub-plugin is TRUE.

If it is FALSE in while loop,  g_cond_wait_until is executed, and when the lock is released, it checks whether the value of eventData is TRUE again.
```
/**
 * @brief Check if current epochs is complete, 
 * tensor_trainer wait for one of epochs to complete before getting the results from the subplugin
 */
static gboolean
gst_tensor_trainer_epochs_is_complete (GstTensorTrainer * trainer)
{
  int required_sample;
  gint64 end_time;
  GstTensorTrainerFrameworkEventData event_data;

  g_return_val_if_fail (trainer != NULL, FALSE);
  g_return_val_if_fail (trainer->fw != NULL, FALSE);
  g_return_val_if_fail (&trainer->prop != NULL, FALSE);

  required_sample =
      trainer->prop.num_training_samples + trainer->prop.num_validation_samples;
  if (trainer->total_push_data_cnt % required_sample != 0)
    return FALSE;

  trainer->fw->eventHandler (trainer->fw, &trainer->prop, trainer->privateData,
      CHECK_EPOCH_COMPLETION, &event_data);

  if (event_data.is_epoch_complete)
    return TRUE;
  GST_ERROR ("event_data.is_epoch_complete (%d)", event_data.is_epoch_complete);
  /* wait for epoch to complete */
  g_mutex_lock (&trainer->epoch_complete_lock);
  while (!event_data.is_epoch_complete) {
    GST_ERROR ("event_data.is_epoch_complete (%d)",
        event_data.is_epoch_complete);
    end_time = g_get_monotonic_time () + 1 * G_TIME_SPAN_SECOND;
    if (!g_cond_wait_until (&trainer->epoch_complete_cond,
            &trainer->epoch_complete_lock, end_time)) {
      /* check again */
      trainer->fw->eventHandler (trainer->fw, &trainer->prop,
          trainer->privateData, CHECK_EPOCH_COMPLETION, &event_data);

      GST_INFO_OBJECT (trainer, "is_epoch_complete(%d)",
          event_data.is_epoch_complete);
    }
  }
  g_mutex_unlock (&trainer->epoch_complete_lock);
  return TRUE;
}
```